File: 2022/day04/python/solutionB.py
#!/bin/env python

input = open('../input.txt').read()

def format(input: str) -> list[list[list[int]]]:
    return [
        [
            [
                int(n) for n in p.split('-')
            ] for p in s.split(',')
        ] for s in input.split('\n')
    ]

def solution(sections: list[list[list[int]]]) -> int:
    def _filter(pair) -> bool:
        first = set(range(pair[0][0], pair[0][1] + 1))
        second = set(range(pair[1][0], pair[1][1] + 1))
        return not first.isdisjoint(second)
    
    return len(list(filter(_filter, sections)))

print(solution(format(input)))

File: 2022/day04/ruby/solutionB.rb
#!/bin/env ruby
# frozen_string_literal: true

require 'set'

input = File.read('../input.txt')

# @param {String} input
# @return {Integer[][][]}
def format_input(input)
  input.split.map do |sections|
    sections.split(',').map do |pair|
      pair.split('-').map(&:to_i)
    end
  end
end

# @param {Integer[][][]} sections
# @return {Integer}
def solution(sections)
  sections.filter do |section|
    first = (section.first[0]..section.first[1])
    second = (section.last[0]..section.last[1])
    !first.to_set.&(second.to_set).empty?
  end.length
end

print solution(format_input(input))